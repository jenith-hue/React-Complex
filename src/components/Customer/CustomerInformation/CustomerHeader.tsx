import React, { useState, useEffect } from 'react';
import {
  makeStyles,
  RACButton,
  RACSelect,
  Typography,
} from '@rentacenter/racstrap';
import clsx from 'clsx';
import {
  API_ERROR_MESSAGE,
  ASSIGN_ROUTE_ERROR_MESSAGE,
  DELETE_ALERT_ERROR_MESSAGE,
  UPDATE_ALERT_ERROR_MESSAGE,
} from '../../../constants/constants';
import {
  useCustomerHeader,
  useCustomerHeaderActions,
} from '../../../context/CustomerHeader/CustomerHeaderProvider';
import { AssignAlertModal } from './AssignAlertModal';
import { CustomerAlert_ as CustomerAlert } from './CustomerAlert';
import { getCustomerFullName } from '../../PastDueList/PastDueCustomerList/CustomerList';
import { ApiErrorModal } from '../../../common/ApiError/ApiError';
import { useCustomerAlerts } from '../../../context/CustomerAlert/CustomerAlertsProvider';
import { useCustomerDetails } from '../../../context/CustomerDetails/CustomerDetailsProvider';
import { useLocation } from 'react-router-dom';
import { CustomerLocationState } from '../../../types/types';

export const CustomerHeaderTestId = 'CustomerHeaderTestId';
export const assignAlertButtonTestId = 'assignAlertButtonTestId';
export const customerNameTestId = 'customerNameTestId';

const useClasses = makeStyles((theme: any) => ({
  customerHeader: {
    display: 'flex',
    flexGrow: 1,
    backgroundColor: 'white',
    padding: theme.typography.pxToRem(8),
    margin: `${theme.typography.pxToRem(0)} -${theme.typography.pxToRem(10)}`,
  },
  marginTop4: {
    marginTop: theme.typography.pxToRem(4),
  },
  marginTop8: {
    marginTop: theme.typography.pxToRem(8),
  },
  customerNameWrapper: { display: 'flex' },
  customerHeaderLabel: {
    fontSize: theme.typography.pxToRem(22),
    lineHeight: theme.typography.pxToRem(26),
    marginTop: theme.typography.pxToRem(8),
    marginBottom: theme.typography.pxToRem(8),
    paddingRight: theme.typography.pxToRem(24),
    width: '50%',
  },
  customerHeaderLeftPanel: {
    display: 'flex',
    minWidth: '41%',
    borderRight: `${theme.typography.pxToRem(1)} solid #D3D3D3`,
  },
  customerName: { color: '#2179FE', display: 'flex' },
  customerHeaderRightPanel: { display: 'flex', minWidth: '61%' },
  alert: {
    width: '58%',
    padding: `${theme.typography.pxToRem(0)} ${theme.typography.pxToRem(16)}`,
  },
  route: {
    display: 'flex',
    width: '42%',
    paddingRight: theme.typography.pxToRem(25),
  },
  assignAlertButtonWrapper: {
    width: '41%',
  },
  assignAlertButton: {
    height: theme.typography.pxToRem(43),
    marginTop: theme.typography.pxToRem(8),
  },
  routeSelect: {
    width: '50%',
    padding: `${theme.typography.pxToRem(0)} ${theme.typography.pxToRem(
      16
    )} ${theme.typography.pxToRem(0)} ${theme.typography.pxToRem(8)}`,
  },
  selectOptionsPaper: {
    maxHeight: theme.typography.pxToRem(200),
  },
}));

export const CustomerHeader = () => {
  const classes = useClasses();
  const location = useLocation<CustomerLocationState>();
  const [openAssignAlertModal, setOpenAssignAlertModal] =
    useState<boolean>(false);

  const {
    routeOptions,
    selectedRouteOption,
    isLoading,
    hasRouteApiError,
    hasAssignRouteApiError,
    isRouteSelectionDisabled,
  } = useCustomerHeader();

  const { onChangeRoute } = useCustomerHeaderActions();

  const [isApiErrorOpen, setIsApiErrorOpen] = useState(false);
  const [apiErrorMessage, setApiErrorMessage] = useState('');
  const {
    customerAlerts,
    hasApiError,
    hasDeleteApiError,
    hasFetchAllAlertsError,
    isFetchAllAlertsLoading,
  } = useCustomerAlerts();
  const activeCustomerAlerts = customerAlerts.filter(
    (alert) => alert.alertCleared === 0
  )?.length;
  const alertsCountLabel = activeCustomerAlerts
    ? `(${activeCustomerAlerts})`
    : '';
  const { customerDetails } = useCustomerDetails();

  const { firstName, lastName } = customerDetails || {};

  const { firstPaymentDefault, secondPaymentDefault } =
    location?.state?.customer || {};

  const handleOnSaveAlert = () => {
    setOpenAssignAlertModal(false);
  };

  useEffect(() => {
    if (hasApiError || hasDeleteApiError || hasAssignRouteApiError) {
      let errorMessage = '';
      if (hasDeleteApiError) {
        errorMessage = DELETE_ALERT_ERROR_MESSAGE;
      } else if (hasAssignRouteApiError) {
        errorMessage = ASSIGN_ROUTE_ERROR_MESSAGE;
      } else {
        errorMessage = UPDATE_ALERT_ERROR_MESSAGE;
      }
      setIsApiErrorOpen(true);
      setApiErrorMessage(errorMessage);
    }
  }, [hasApiError, hasDeleteApiError, hasAssignRouteApiError]);

  return (
    <>
      {isApiErrorOpen && (
        <ApiErrorModal
          open={isApiErrorOpen}
          onClose={() => setIsApiErrorOpen(false)}
          message={apiErrorMessage}
        />
      )}
      <div
        className={classes.customerHeader}
        data-testid={CustomerHeaderTestId}
      >
        <div
          className={clsx(classes.customerHeaderLeftPanel, classes.marginTop8)}
        >
          {openAssignAlertModal && (
            <AssignAlertModal
              onSave={handleOnSaveAlert}
              onClose={() => setOpenAssignAlertModal(false)}
              open={openAssignAlertModal}
            />
          )}
          <Typography
            variant="h5"
            component="h2"
            classes={{ root: classes.customerHeaderLabel }}
          >
            Customer Information
          </Typography>
          <div className={classes.marginTop4}>
            <Typography display="inline" variant="body1">
              Customer Name
            </Typography>
            <div
              className={clsx(classes.customerNameWrapper, classes.marginTop4)}
            >
              <div
                className={clsx(classes.customerName, classes.marginTop4)}
                data-testid={customerNameTestId}
              >
                <Typography display="inline" variant="body1">
                  {getCustomerFullName(
                    firstName || '',
                    lastName || '',
                    firstPaymentDefault || '',
                    secondPaymentDefault || ''
                  )}
                </Typography>
              </div>
            </div>
          </div>
        </div>
        <div
          className={clsx(classes.customerHeaderRightPanel, classes.marginTop8)}
        >
          <div className={classes.alert}>
            <Typography display="inline" variant="body1">
              Customer Alerts {alertsCountLabel}
            </Typography>
            <CustomerAlert />
          </div>
          <div className={classes.route}>
            <div className={classes.assignAlertButtonWrapper}>
              <RACButton
                disabled={isFetchAllAlertsLoading || hasFetchAllAlertsError}
                variant="contained"
                size="large"
                color="primary"
                className={classes.assignAlertButton}
                data-testid={assignAlertButtonTestId}
                onClick={() => {
                  setOpenAssignAlertModal(true);
                }}
              >
                Assign Alert
              </RACButton>
            </div>
            <div className={classes.routeSelect}>
              <RACSelect
                classes={{ paper: classes.selectOptionsPaper }}
                inputLabel="Route"
                defaultValue={selectedRouteOption + ''}
                options={routeOptions}
                loading={isLoading}
                isDisabled={isRouteSelectionDisabled}
                onChange={(e: React.ChangeEvent<HTMLSelectElement>) =>
                  onChangeRoute(e.target.value)
                }
                {...(hasRouteApiError && {
                  errorMessage: API_ERROR_MESSAGE,
                })}
              />
            </div>
          </div>
        </div>
      </div>
    </>
  );
};
