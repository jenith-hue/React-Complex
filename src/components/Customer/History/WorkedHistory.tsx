import {
  makeStyles,
  RACCOLOR,
  RACRowTip,
  RACTable,
  RACTableCell,
  RACTableRow,
  Typography,
} from '@rentacenter/racstrap';
import React, { useContext, useEffect, useState } from 'react';
import InfiniteScroll from 'react-infinite-scroll-component';
import {
  ApiStateWrapper,
  Loader,
} from '../../../common/ApiStateWrapper/ApiStateWrapper';
import { LegendItemList } from '../../../common/Legend/LegendItemList';
import {
  INITIAL_OFFSET,
  WORKED_HISTORY_ACTIVITY_TYPE_CODES,
  WORKED_HISTORY_API_LIMIT,
} from '../../../constants/constants';
import { useCustomerDetails } from '../../../context/CustomerDetails/CustomerDetailsProvider';
import {
  PastDueListCustomerStateContext,
  PastDueListCustomerDispatchStateContext,
} from '../../../context/PastDueListCustomer/PastDueListCustomerProvider';
import {
  useWorkedHistory,
  useWorkedHistoryActions,
} from '../../../context/WorkedHistory/WorkedHistoryProvider';
import { WorkedHistoryStatusLegendNames } from '../../../domain/WorkedHistory/WorkedHistory';
import { WorkedHistoryCommitmentType } from '../../../types/types';
import {
  abbrev,
  formatDateString,
  formatPhoneNumber,
  formatStringDateHoursAndMinutes,
} from '../../../utils/utils';
import { WorkedHistoryNotesModal } from './WorkedHistoryNotesModal';
import { CommitmentNotesModal } from './CommitmentNotesModal';

const useClasses = makeStyles((theme: any) => ({
  legendWrapper: {
    height: theme.typography.pxToRem(70),
  },
  tableContentColor: {
    color: RACCOLOR.INDEPENDENCE,
  },
  noLeftPadding: {
    paddingLeft: 0,
  },
  tableRoot: {
    marginTop: theme.typography.pxToRem(16),
  },
  tableHead: {
    position: 'sticky',
    top: 0,
    zIndex: 2,
    background: 'white',
  },
  contentHeight: {
    height: theme.typography.pxToRem(330),
  },
  // row: {
  //   display: 'flex',
  // },
  notes: {
    cursor: 'pointer',
  },
  tipRoot: {
    borderLeftWidth: theme.typography.pxToRem(6),
  },
}));

export const WorkedHistory = () => {
  const [offset, setOffset] = useState<number>(INITIAL_OFFSET);
  const [isFirstFetch, setIsFirstFetch] = useState<boolean>(true);
  const {
    workedHistory: data,
    hasApiError,
    loading,
    reloadWorkedHistory,
    hasMore,
  } = useWorkedHistory();
  const { filter } = useContext(PastDueListCustomerStateContext);

  const { isLogActivityAllowed, customerDetails } = useCustomerDetails();
  const { firstName, lastName } = customerDetails || {};

  const { setSeletedFilter } = useContext(
    PastDueListCustomerDispatchStateContext
  );

  const onSelect = (selectedFilter: string) => {
    filter === selectedFilter
      ? setSeletedFilter('')
      : setSeletedFilter(selectedFilter);
  };
  const [isNotesModalOpen, setIsNotesModalOpen] = useState(false);
  const [isCommitmentNotesModalOpen, setIsCommitmentNotesModalOpen] =
    useState(false);
  const [selectedCommitment, setSelectedCommitment] =
    useState<WorkedHistoryCommitmentType | null>();
  const [workedResultText, setWorkedResultText] = useState('');
  const [noteText, setNoteText] = useState('');

  const handleCommitmentNotesModalClose = () => {
    setIsCommitmentNotesModalOpen(false);
    setSelectedCommitment(null);
  };

  const renderTableHead = () => (
    <>
      <RACTableCell>
        <Typography variant="h5">Worked Date & Time</Typography>
      </RACTableCell>
      <RACTableCell>
        <Typography variant="h5">Worked Result</Typography>
      </RACTableCell>
      <RACTableCell>
        <Typography variant="h5">Notes</Typography>
      </RACTableCell>
      <RACTableCell>
        <Typography variant="h5">Commitment Status</Typography>
      </RACTableCell>
      <RACTableCell>
        <Typography variant="h5">Due Date</Typography>
      </RACTableCell>
      <RACTableCell align="center">
        <Typography variant="h5">Days Late</Typography>
      </RACTableCell>
      <RACTableCell>
        <Typography variant="h5">Phone # Dialed</Typography>
      </RACTableCell>
      <RACTableCell>
        <Typography variant="h5">Relation</Typography>
      </RACTableCell>
      <RACTableCell>
        <Typography variant="h5">Source</Typography>
      </RACTableCell>
    </>
  );
  const renderTableContent = () => (
    <>
      {data?.map((dataObj, index) => {
        const code: string | undefined = dataObj.accountActivityType?.code;
        const description: string | undefined =
          code?.toUpperCase() ===
          WORKED_HISTORY_ACTIVITY_TYPE_CODES.CALL_CUSTOMER
            ? dataObj.callResultType?.description
            : dataObj.accountActivityType?.description;
        const workedDateString = `${formatDateString(
          dataObj.activityDate
        )} ${formatStringDateHoursAndMinutes(dataObj.activityDate)}`;

        const noteText = dataObj.commitment?.notes || dataObj?.notes || '';

        const soureName: string | undefined =
          code?.toUpperCase() ===
          WORKED_HISTORY_ACTIVITY_TYPE_CODES.TEXT_RECEIVED
            ? `${firstName} ${lastName}`
            : `${dataObj.coWorker?.firstName || ''} ${
                dataObj.coWorker?.lastName || ''
              }`;

        return (
          <RACTableRow key={index} hover backgroundColor={dataObj.color?.row}>
            <RACTableCell
              classes={{
                root: classes.noLeftPadding,
              }}
            >
              <div
              // className={classes.row}
              >
                <RACRowTip
                  classes={{ root: classes.tipRoot }}
                  tipColor={dataObj.color?.tip || RACCOLOR.WHITE}
                />
                <Typography
                  display="inline"
                  variant="body2"
                  className={classes.tableContentColor}
                >
                  {workedDateString}
                </Typography>
              </div>
            </RACTableCell>
            <RACTableCell>
              <Typography variant="body2">{description}</Typography>
            </RACTableCell>
            <RACTableCell>
              <Typography
                noWrap
                variant="body2"
                className={classes.tableContentColor}
              >
                <Typography noWrap variant="body2" color="primary">
                  <span
                    className={classes.notes}
                    onClick={() => {
                      if (dataObj.commitment) {
                        setSelectedCommitment(dataObj.commitment);
                        setIsCommitmentNotesModalOpen(true);
                      } else {
                        setIsNotesModalOpen(true);
                        setWorkedResultText(description || '');
                        setNoteText(noteText);
                      }
                    }}
                  >
                    {abbrev(noteText)}
                  </span>
                </Typography>
              </Typography>
            </RACTableCell>
            <RACTableCell>
              <Typography variant="body2" className={classes.tableContentColor}>
                {dataObj.commitment?.commitmentStatus?.description}
              </Typography>
            </RACTableCell>
            <RACTableCell>
              <Typography variant="body2" className={classes.tableContentColor}>
                {formatDateString(dataObj.pastDueDate)}
              </Typography>
            </RACTableCell>
            <RACTableCell>
              <Typography
                align="center"
                variant="body2"
                className={classes.tableContentColor}
              >
                {dataObj.daysLate}
              </Typography>
            </RACTableCell>
            <RACTableCell>
              <Typography variant="body2" className={classes.tableContentColor}>
                {formatPhoneNumber(
                  dataObj.phoneNumberDialed || '',
                  !isLogActivityAllowed
                )}
              </Typography>
            </RACTableCell>
            <RACTableCell>
              <Typography variant="body2" className={classes.tableContentColor}>
                {dataObj.phoneContact?.relationshipType?.description || ''}
              </Typography>
            </RACTableCell>
            <RACTableCell>
              <Typography variant="body2" className={classes.tableContentColor}>
                {soureName}
              </Typography>
            </RACTableCell>
          </RACTableRow>
        );
      })}
    </>
  );

  const classes = useClasses();
  const { fetchWorkedHistory } = useWorkedHistoryActions();

  useEffect(() => {
    if (!isFirstFetch && reloadWorkedHistory) {
      fetchWorkedHistory(offset, WORKED_HISTORY_API_LIMIT);
    }
    setIsFirstFetch(false);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [reloadWorkedHistory]);

  useEffect(() => {
    if (offset === INITIAL_OFFSET && isFirstFetch) {
      fetchWorkedHistory(offset, WORKED_HISTORY_API_LIMIT);
      setIsFirstFetch(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (!isFirstFetch) {
      fetchWorkedHistory(offset, WORKED_HISTORY_API_LIMIT);
    }
    setIsFirstFetch(false);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [offset]);

  /*
  Increment the offset.
  Fetch is handled in the useEffect that listens on the offset.
  */
  const fetchNextPage = () => {
    /*
    fetchNextPage is used by infinite scroll component
    the first fetch is handled by our component, therefore we are not letting the
    infinite scroll to trigger the first request.
    Handles the case when you apply the filters
    (offset set to 0 -> request is triggered, but infinite scroll is also triggering a request, because its data length is 0)
    */
    data.length > 1 && setOffset(offset + WORKED_HISTORY_API_LIMIT);
  };

  return (
    <>
      {isNotesModalOpen && (
        <WorkedHistoryNotesModal
          noteText={noteText}
          workedResultText={workedResultText}
          open={isNotesModalOpen}
          onClose={() => setIsNotesModalOpen(false)}
        />
      )}
      {isCommitmentNotesModalOpen && selectedCommitment && (
        <CommitmentNotesModal
          commitment={selectedCommitment}
          onClose={handleCommitmentNotesModalClose}
        />
      )}
      <div className={classes.legendWrapper}>
        <LegendItemList
          list={WorkedHistoryStatusLegendNames}
          onSelect={onSelect}
        />
      </div>
      <InfiniteScroll
        next={fetchNextPage}
        dataLength={data.length}
        hasMore={hasMore}
        height={330}
        loader={loading && data?.length ? <Loader /> : null}
      >
        <ApiStateWrapper
          loading={loading && data?.length === 0}
          hasApiError={hasApiError && data?.length === 0}
          response={data}
          successContent={
            <RACTable
              headClasses={{ root: classes.tableHead }}
              classes={{ root: classes.tableRoot }}
              renderTableHead={renderTableHead}
              renderTableContent={renderTableContent}
            />
          }
          classes={{
            loader: classes.contentHeight,
            apiError: classes.contentHeight,
            noItems: classes.contentHeight,
          }}
        />
      </InfiniteScroll>
    </>
  );
};
