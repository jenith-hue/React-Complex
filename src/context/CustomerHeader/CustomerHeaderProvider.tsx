import { CancelTokenSource } from 'axios';
import React, {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState,
} from 'react';
import { getCancelTokenSource } from '../../api/client';
import { assignRoute } from '../../api/Customer';
import { ALL_OPTION } from '../../constants/constants';
import { AssignRoutePayload, RouteOption } from '../../types/types';
import { getSelectedStore } from '../../utils/utils';
import { useCustomerDetails } from '../CustomerDetails/CustomerDetailsProvider';
import { PastDueListCustomerDispatchStateContext } from '../PastDueListCustomer/PastDueListCustomerProvider';
import { useFilters } from '../PastDueListSearchCriteria/PastDueListSearchCriteriaProvider';
import { useUserPermissions } from '../permission/PermissionsProvider';

export interface CustomerHeaderState {
  routeOptions: RouteOption[];
  selectedRouteOption?: string;
  hasRouteApiError: boolean;
  hasAssignRouteApiError: boolean;
  isRouteSelectionDisabled: boolean;
  isLoading: boolean;
}

export interface CustomerHeaderDispatchState {
  onChangeRoute: (value: string) => void;
  onCloseAssignRouteErrorModal: () => void;
}

export const CustomerHeaderStateContext = createContext<CustomerHeaderState>(
  {} as CustomerHeaderState
);
export const CustomerHeaderDispatchContext =
  createContext<CustomerHeaderDispatchState>({} as CustomerHeaderDispatchState);

export const CustomerHeaderProvider = (props: { children: ReactNode }) => {
  const { customerDetails } = useCustomerDetails();

  const { customerRoutes, customerId } = customerDetails || {};
  const { routeCode } = customerRoutes?.[0] || {};
  const [routeOptions, setRouteOptions] = useState<RouteOption[]>([]);
  const [selectedRouteOption, setSelectedRouteOption] = useState<string>(
    routeCode || ''
  );
  const [hasAssignRouteApiError, setHasAssignRouteApiError] =
    useState<boolean>(false);
  const [isRouteSelectionDisabled, setIsRouteSelectionDisabled] =
    useState<boolean>(true);
  const [isLoading, setIsLoading] = useState<boolean>(false);

  const { canChangeCustomerRoute } = useUserPermissions();

  const { onRouteChanged } = useContext(
    PastDueListCustomerDispatchStateContext
  );

  const {
    isFetchRoutesLoading,
    routeOptions: routeOptionsFromFilters,
    hasRouteApiError,
  } = useFilters();

  useEffect(() => {
    setIsLoading(isFetchRoutesLoading);
  }, [isFetchRoutesLoading]);

  useEffect(() => {
    setRouteOptions(
      routeOptionsFromFilters.filter((option) => option.value !== ALL_OPTION)
    );
  }, [routeOptionsFromFilters]);

  useEffect(() => {
    setIsRouteSelectionDisabled(!canChangeCustomerRoute);

    const cancelToken: CancelTokenSource = getCancelTokenSource();
    return () => {
      cancelToken.cancel();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    setIsRouteSelectionDisabled(!canChangeCustomerRoute);
  }, [canChangeCustomerRoute]);

  const onChangeRoute = async (routeCode: string) => {
    try {
      if (customerId) {
        const payload = {
          storeNumber: getSelectedStore(),
          routeRefCode: routeCode,
          customerId: Number(customerId),
        } as AssignRoutePayload;

        setHasAssignRouteApiError(false);
        setIsLoading(true);

        await assignRoute(payload);
        const route = routeOptions.find(
          (route) => route.routeCode === routeCode
        );
        setSelectedRouteOption(route?.value || route?.routeCode || '');

        onRouteChanged(route, customerId || '');
      }
    } catch {
      setHasAssignRouteApiError(true);
    } finally {
      setIsLoading(false);
    }
  };

  const onCloseAssignRouteErrorModal = () => {
    setHasAssignRouteApiError(false);
  };
  return (
    <CustomerHeaderStateContext.Provider
      value={{
        routeOptions,
        selectedRouteOption,
        hasRouteApiError,
        hasAssignRouteApiError,
        isRouteSelectionDisabled,
        isLoading,
      }}
    >
      <CustomerHeaderDispatchContext.Provider
        value={{
          onChangeRoute,
          onCloseAssignRouteErrorModal,
        }}
      >
        {props.children}
      </CustomerHeaderDispatchContext.Provider>
    </CustomerHeaderStateContext.Provider>
  );
};

export const useCustomerHeader = () => useContext(CustomerHeaderStateContext);

export const useCustomerHeaderActions = () =>
  useContext(CustomerHeaderDispatchContext);
