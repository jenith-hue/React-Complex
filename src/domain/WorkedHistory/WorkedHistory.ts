import { RACCOLOR } from '@rentacenter/racstrap';

export enum WorkedHistoryStatusLegend {
  LeftMessage = 'LeftMessage',
  Commitment = 'Commitment',
  NoContact = 'NoContact',
  BrokenCommitment = 'BrokenCommitment',
  TextSent = 'TextSent',
  TextReceived = 'TextReceived',
}

export interface WorkedHistoryStatusLegendEntityProps {
  label: string;
  color?: string;
}

export const WorkedHistoryStatusLegendNames: Partial<
  Record<WorkedHistoryStatusLegend, WorkedHistoryStatusLegendEntityProps>
> = {
  [WorkedHistoryStatusLegend.LeftMessage]: {
    label: 'Left Message',
    color: RACCOLOR.TUFTS_BLUE,
  },
  [WorkedHistoryStatusLegend.Commitment]: {
    label: 'Commitment',
    color: RACCOLOR.HONEYCOMB,
  },
  [WorkedHistoryStatusLegend.NoContact]: {
    label: 'No Contact',
    color: RACCOLOR.RICH_LILAC,
  },
  [WorkedHistoryStatusLegend.BrokenCommitment]: {
    label: 'Broken Commitment',
    color: RACCOLOR.PUNCH,
  },
  [WorkedHistoryStatusLegend.TextSent]: {
    label: 'Text Sent',
    color: RACCOLOR.BLUE_MANA,
  },
  [WorkedHistoryStatusLegend.TextReceived]: {
    label: 'Text Received',
    color: RACCOLOR.INTERGALACTIC_COWBOY,
  },
};

export interface LogWorkedHistoryPayload {
  storeNumber: string;
  activityDate: string;
  notes: string;
  callResultRefCode: string;
  commitmentId: string;
  acctActivityRefCode: string;
  daysPastDue: string;
  customerId: string | number;
  phoneNumberDialed: string;
  coWorkerId: string;
  receiptId: string;
  callTime: string;
}
