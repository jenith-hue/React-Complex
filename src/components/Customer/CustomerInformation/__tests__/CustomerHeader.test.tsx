/* eslint-disable react/display-name */

import React from 'react';
import renderer from 'react-test-renderer';
import { CustomerHeader } from '../CustomerHeader';
import * as workedHistory from '../../../../context/WorkedHistory/WorkedHistoryProvider';

interface PropWithChildren {
  children: JSX.Element;
}

jest.mock('@rentacenter/racstrap', () => ({
  Typography: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="Typography" {...rest}>
      {children}
    </div>
  ),
  RACButton: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACButton" {...rest}>
      {children}
    </div>
  ),
  RACSelect: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACCard" {...rest}>
      {children}
    </div>
  ),
  RACCOLOR: {
    WHITE: 'WHITE',
  },
  makeStyles: () => () => ({}),
}));

jest.mock('../../../../context/CustomerHeader/CustomerHeaderProvider', () => ({
  useCustomerHeaderActions: () => ({
    onChangeRoute: jest.fn(),
    onCloseAssignRouteErrorModal: jest.fn(),
  }),
  useCustomerHeader: () => ({
    routeOptions: [],
    selectedRouteOption: '',
    isLoading: false,
    hasRouteApiError: false,
    hasAssignRouteApiError: false,
    isRouteSelectionDisabled: false,
  }),
}));

jest.mock('../../../../context/CustomerAlert/CustomerAlertsProvider', () => ({
  useCustomerAlerts: () => ({
    customerAlerts: [],
    loading: false,
    hasApiError: false,
    hasDeleteApiError: false,
    hasGetApiError: false,
  }),
  useCustomerAlertsActions: () => ({
    fetchCustomerAlerts: jest.fn(),
    removeCustomerAlert: jest.fn(),
    saveCustomerAlerts: jest.fn(),
  }),
}));
// Notice that the below is required, even if there is an another
// jest.spyOn in the test.
// Otherwise, we'll get the following: TypeError: Cannot redefine property: useLocation
jest.mock('react-router', () => ({
  ...jest.requireActual('react-router'),
  useLocation: () => ({
    pathname: '',
    state: {
      customer: {
        customerFirstName: 'John',
        customerLastName: 'Doe',
        firstPaymentDefault: 'Y',
        secondPaymentDefault: 'N',
      },
    },
  }),
}));
describe('CustomerHeader', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<CustomerHeader />);

    expect(tree.toJSON()).toMatchSnapshot();
  });
  it('renders correctly, when worked history label is present', () => {
    jest
      .spyOn(workedHistory, 'useWorkedHistory')
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      //@ts-ignore
      .mockImplementationOnce(() => ({
        workedHistoryLabel: {
          text: 'Text',
          background: 'BLACK',
        },
      }));
    const tree = renderer.create(<CustomerHeader />);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
