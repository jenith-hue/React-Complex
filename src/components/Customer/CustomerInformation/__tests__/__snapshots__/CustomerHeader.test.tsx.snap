// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerHeader renders correctly 1`] = `
<div
  data-testid="CustomerHeaderTestId"
>
  <div
    className=""
  >
    <div
      classes={
        Object {
          "root": undefined,
        }
      }
      component="h2"
      data-testid="Typography"
      variant="h5"
    >
      Customer Information
    </div>
    <div>
      <div
        data-testid="Typography"
        display="inline"
        variant="body1"
      >
        Customer Name
      </div>
      <div
        className=""
      >
        <div
          className=""
          data-testid="customerNameTestId"
        >
          <div
            data-testid="Typography"
            display="inline"
            variant="body1"
          >
            * 
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className=""
  >
    <div>
      <div
        data-testid="Typography"
        display="inline"
        variant="body1"
      >
        Customer Alerts 
        
      </div>
      <div
        data-testid="customerAlertTestId"
      />
    </div>
    <div>
      <div>
        <div
          color="primary"
          data-testid="assignAlertButtonTestId"
          onClick={[Function]}
          size="large"
          variant="contained"
        >
          Assign Alert
        </div>
      </div>
      <div>
        <div
          classes={
            Object {
              "paper": undefined,
            }
          }
          data-testid="RACCard"
          defaultValue=""
          inputLabel="Route"
          isDisabled={false}
          loading={false}
          onChange={[Function]}
          options={Array []}
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`CustomerHeader renders correctly, when worked history label is present 1`] = `
<div
  data-testid="CustomerHeaderTestId"
>
  <div
    className=""
  >
    <div
      classes={
        Object {
          "root": undefined,
        }
      }
      component="h2"
      data-testid="Typography"
      variant="h5"
    >
      Customer Information
    </div>
    <div>
      <div
        data-testid="Typography"
        display="inline"
        variant="body1"
      >
        Customer Name
      </div>
      <div
        className=""
      >
        <div
          className=""
          data-testid="customerNameTestId"
        >
          <div
            data-testid="Typography"
            display="inline"
            variant="body1"
          >
            * 
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className=""
  >
    <div>
      <div
        data-testid="Typography"
        display="inline"
        variant="body1"
      >
        Customer Alerts 
        
      </div>
      <div
        data-testid="customerAlertTestId"
      />
    </div>
    <div>
      <div>
        <div
          color="primary"
          data-testid="assignAlertButtonTestId"
          onClick={[Function]}
          size="large"
          variant="contained"
        >
          Assign Alert
        </div>
      </div>
      <div>
        <div
          classes={
            Object {
              "paper": undefined,
            }
          }
          data-testid="RACCard"
          defaultValue=""
          inputLabel="Route"
          isDisabled={false}
          loading={false}
          onChange={[Function]}
          options={Array []}
        />
      </div>
    </div>
  </div>
</div>
`;
